File: apps/api/app/api/v1/admin/categories.py
"""Admin category endpoints. Require role='admin'; non-admins see 404."""

from __future__ import annotations

from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query

from app.api.deps import DbSession, require_role
from app.models.user import User
from app.schemas.admin_catalog import (
    CategoryAdmin,
    CategoryAdminListOut,
    CategoryAdminOut,
    CategoryCreateIn,
    CategoryUpdateIn,
)
from app.services.catalog import admin as admin_service

router = APIRouter(
    prefix="/admin/categories",
    tags=["admin-catalog"],
    dependencies=[Depends(require_role("admin"))],
)


def _serialize(cat) -> CategoryAdmin:  # type: ignore[no-untyped-def]
    return CategoryAdmin(
        id=cat.id,
        name=cat.name,
        slug=cat.slug,
        parent_id=cat.parent_id,
        description=cat.description,
        image_url=cat.image_url,
        sort_order=cat.sort_order,
        status=cat.status,
        created_at=cat.created_at,
        updated_at=cat.updated_at,
        deleted_at=cat.deleted_at,
    )


@router.get("", response_model=CategoryAdminListOut)
async def list_categories(
    db: DbSession,
    status: Annotated[Literal["active", "archived"] | None, Query()] = None,
) -> CategoryAdminListOut:
    cats = await admin_service.list_categories(db, status=status)
    return CategoryAdminListOut(data=[_serialize(c) for c in cats])


@router.post("", response_model=CategoryAdminOut, status_code=201)
async def create_category(
    payload: CategoryCreateIn,
    db: DbSession,
    actor: Annotated[User, Depends(require_role("admin"))],
) -> CategoryAdminOut:
    cat = await admin_service.create_category(
        db,
        actor=actor,
        name=payload.name,
        slug=payload.slug,
        parent_id=payload.parent_id,
        description=payload.description,
        image_url=payload.image_url,
        sort_order=payload.sort_order,
        status=payload.status,
    )
    return CategoryAdminOut(category=_serialize(cat))


@router.get("/{category_id}", response_model=CategoryAdminOut)
async def get_category(category_id: str, db: DbSession) -> CategoryAdminOut:
    cat = await admin_service.get_category(db, category_id)
    return CategoryAdminOut(category=_serialize(cat))


@router.patch("/{category_id}", response_model=CategoryAdminOut)
async def update_category(
    category_id: str,
    payload: CategoryUpdateIn,
    db: DbSession,
    actor: Annotated[User, Depends(require_role("admin"))],
) -> CategoryAdminOut:
    updates = payload.model_dump(exclude_unset=True)
    cat = await admin_service.update_category(
        db, actor=actor, category_id=category_id, updates=updates
    )
    return CategoryAdminOut(category=_serialize(cat))


@router.delete("/{category_id}", status_code=204)
async def delete_category(
    category_id: str,
    db: DbSession,
    actor: Annotated[User, Depends(require_role("admin"))],
) -> None:
    await admin_service.delete_category(db, actor=actor, category_id=category_id)

File: apps/web/src/services/categories.ts
import type {
  Category,
  CategoryCreatePayload,
  CategoryStatus,
  CategoryUpdatePayload,
} from '@/types/catalog';

import { apiFetch } from './api/client';

interface CategoryApi {
  id: string;
  name: string;
  slug: string;
  parent_id: string | null;
  description: string | null;
  image_url: string | null;
  sort_order: number;
  status: CategoryStatus;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

function fromApi(c: CategoryApi): Category {
  return {
    id: c.id,
    name: c.name,
    slug: c.slug,
    parentId: c.parent_id,
    description: c.description,
    imageUrl: c.image_url,
    sortOrder: c.sort_order,
    status: c.status,
    createdAt: c.created_at,
    updatedAt: c.updated_at,
    deletedAt: c.deleted_at,
  };
}

export async function listCategories(opts?: {
  status?: CategoryStatus;
}): Promise<Category[]> {
  const qs = opts?.status ? `?status=${opts.status}` : '';
  const data = await apiFetch<{ data: CategoryApi[] }>(`/admin/categories${qs}`);
  return data.data.map(fromApi);
}

export async function getCategory(id: string): Promise<Category> {
  const data = await apiFetch<{ category: CategoryApi }>(`/admin/categories/${id}`);
  return fromApi(data.category);
}

export async function createCategory(
  payload: CategoryCreatePayload,
): Promise<Category> {
  const data = await apiFetch<{ category: CategoryApi }>('/admin/categories', {
    method: 'POST',
    body: JSON.stringify({
      name: payload.name,
      slug: payload.slug,
      parent_id: payload.parentId ?? null,
      description: payload.description ?? null,
      image_url: payload.imageUrl ?? null,
      sort_order: payload.sortOrder ?? 0,
      status: payload.status ?? 'active',
    }),
  });
  return fromApi(data.category);
}

export async function updateCategory(
  id: string,
  payload: CategoryUpdatePayload,
): Promise<Category> {
  const body: Record<string, unknown> = {};
  if (payload.name !== undefined) body.name = payload.name;
  if (payload.slug !== undefined) body.slug = payload.slug;
  if (payload.parentId !== undefined) body.parent_id = payload.parentId;
  if (payload.description !== undefined) body.description = payload.description;
  if (payload.imageUrl !== undefined) body.image_url = payload.imageUrl;
  if (payload.sortOrder !== undefined) body.sort_order = payload.sortOrder;
  if (payload.status !== undefined) body.status = payload.status;

  const data = await apiFetch<{ category: CategoryApi }>(
    `/admin/categories/${id}`,
    {
      method: 'PATCH',
      body: JSON.stringify(body),
    },
  );
  return fromApi(data.category);
}

export async function deleteCategory(id: string): Promise<void> {
  await apiFetch<void>(`/admin/categories/${id}`, { method: 'DELETE' });
}
